core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -282,15 +282,15 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
   /**
    * Update high watermark with offset metadata. The new high watermark will be lower
-   * bounded by the log start offset and upper bounded by the log end offset.
+   * bounded by the local-log-start-offset and upper bounded by the log-end-offset.
    *
    * @param highWatermarkMetadata the suggested high watermark with offset metadata
    * @return the updated high watermark offset
    */
   def updateHighWatermark(highWatermarkMetadata: LogOffsetMetadata): Long = {
     val endOffsetMetadata = localLog.logEndOffsetMetadata
-    val newHighWatermarkMetadata = if (highWatermarkMetadata.messageOffset < logStartOffset) {
-      new LogOffsetMetadata(logStartOffset)
+    val newHighWatermarkMetadata = if (highWatermarkMetadata.messageOffset < _localLogStartOffset) {
```

Comment:
Hmm, when will we set HWM to be lower than _localLogStartOffset?

In `UnifiedLog.deletableSegments()`, we have the following code that bounds the retention based deletion by highWatermark. When updating highWatermark, the value typically increases.
`        val predicateResult = highWatermark >= upperBoundOffset && predicate(segment, nextSegmentOpt)
`

Reply:
> when will we set HWM to be lower than _localLogStartOffset?

This can happen when recovering the partition due to ungraceful shutdown and the replication-offset-checkpoint file is missing/corrupted. When the broker comes online, HWM is set to to localLogStartOffset in [UnifiedLog#updateLocalLogStartOffset](https://sourcegraph.com/github.com/apache/kafka@f895ab5145077c5efa10a4a898628d901b01e2c2/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L162), then we load the HWM from the checkpoint file in [Partition#createLog](https://sourcegraph.com/github.com/apache/kafka@f895ab5145077c5efa10a4a898628d901b01e2c2/-/blob/core/src/main/scala/kafka/cluster/Partition.scala?L495).

If the HWM checkpoint file is missing / does not contain the entry for partition, then the default value of 0 is taken. If 0 < LogStartOffset (LSO), then LSO is assumed as HWM . Thus, the non-monotonic update of highwatermark from LLSO to LSO can happen.

Reply:
> In the rare case, the restarted broker is elected as the leader before caught up through unclean election. Is this the case that you want to address?

yes, we want to address this case too. And, the issue can also happen during clean preferred-leader-election:

```
Call stack: The replica (1002) has full data but HW is invalid, then the fetch-offset will be equal to LeaderLog(1001).highWatermark

Leader (1001):
    KafkaApis.handleFetchRequest
        ReplicaManager.fetchMessages
            ReplicaManager.readFromLocalLog
                Partition.fetchRecords
                    Partition.updateFollowerFetchState
                        Partition.maybeExpandIsr
                            Partition.submitAlterPartition
                            ...
                            ...
                            ...
        # If there is not enough data to respond and there is no remote data, we will let the fetch request wait for new data.
        # parks the request in the DelayedFetchPurgatory


Another thread, runs Preferred-Leader-Election in controller (1003), since the replica 1002 joined the ISR list, it can be elected as the preferred leader. The controller sends LeaderAndIsr requests to all the brokers.

    KafkaController.processReplicaLeaderElection
        KafkaController.onReplicaElection
            PartitionStateMachine.handleStateChanges
                PartitionStateMachine.doHandleStateChanges
                    PartitionStateMachine.electLeaderForPartitions
                ControllerChannelManager.sendRequestsToBrokers


Replica 1002 got elected as Leader and have invalid highWatermark since it didn't processed the fetch-response from the previous leader 1001, throws OFFSET_OUT_OF_RANGE error when processing the LeaderAndIsr request. Note that if the LeaderAndIsr request fails to process for one partition, then the remaining partitions in that request won't be processed.

    KafkaApis.handleLeaderAndIsrRequest
        ReplicaManager.becomeLeaderOrFollower
            ReplicaManager.makeLeaders
                Partition.makeLeader
                    Partition.maybeIncrementLeaderHW
                        UnifiedLog.maybeIncrementHighWatermark (LeaderLog)
                            UnifiedLog.fetchHighWatermarkMetadata


The controller assumes that the current-leader for the tp0 is 1002, but the broker 1002 couldn't process the LISR. The controller retries the LISR until the broker becomes leader for tp0. During this time, the producers won't be able to send messages, as the node 1002, sends NOT_LEADER_FOR_PARTITION errorcode to the producer.


```

Reply:
> For the makeLeaders path, it will call UnifiedLog.convertToOffsetMetadataOrThrow. Within it, checkLogStartOffset(offset) shouldn't throw OFFSET_OUT_OF_RANGE since we are comparing the offset with logStartOffset. Do you know which part throws OFFSET_OUT_OF_RANGE error?

The next line `localLog.convertToOffsetMetadataOrThrow` in [convertToOffsetMetadataOrThrow](https://sourcegraph.com/github.com/apache/kafka@a3dcbd4e28a35f79f75ec1bf316ef0b39c0df164/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L1429) method reads the segment from disk, there it throws the error. The call [segments.floorSegment(startOffset)](https://sourcegraph.com/github.com/apache/kafka@a3dcbd4e28a35f79f75ec1bf316ef0b39c0df164/-/blob/core/src/main/scala/kafka/log/LocalLog.scala?L366) in LocalLog fails to find the segment with log-start-offset, then OffsetOutOfRangeException is thrown.

> For the follower fetch path, it's bounded by LogEndOffset. So it shouldn't need to call UnifiedLog.fetchHighWatermarkMetadata, right? The regular consumer will call UnifiedLog.fetchHighWatermarkMetadata.

yes, you're right. I attached the wrong call stack for handling the follower request. Please find the updated call stack below:

Leader with invalid high-watermark handles the FETCH requests from follower and throws OFFSET_OUT_OF_RANGE error:

```
 KafkaApis.handleFetchRequest
        ReplicaManager.fetchMessages
            ReplicaManager.readFromLog
                Partition.fetchRecords
                    # readFromLocalLog
                    Partition.updateFollowerFetchState
                        Partition.maybeIncrementLeaderHW
                             LeaderLog.maybeIncrementHighWatermark
                                UnifiedLog.fetchHighWatermarkMetadata
                                    UnifiedLog.convertToOffsetMetadataOrThrow
                                        LocalLog.convertToOffsetMetadataOrThrow
                                            LocalLog.read
                                                # OffsetOutOfRangeException exception

```

Reply:
@kamalcph : Thanks for the explanation. I understand the problem now.

As for the fix, it seems that it could work for HWM. However, I am not sure that we could always do the same thing of LastStableOffset. For example, if we lose the local data in all replicas, the lastStableOffset could still be in the middle of a tiered segment and moving it to localLogStartOffset immediately will be incorrect. 

Here is another potential approach. Note that OffsetMetadata (segmentBaseOffset and relativePositionInSegment) is only used in DelayedFetch for estimating the amount of available bytes. If occasionally OffsetMetadata is not available, we don't have to force an exception in convertToOffsetMetadataOrThrow(). Instead, we can leave the OffsetMetadata as empty and just use a conservative 1 byte for estimating the amount of available bytes. This approach will apply to both HWM and LSO.  The inaccurate byte estimate will be ok as long as it's infrequent. What do you think?

Reply:
Thanks for suggesting the alternative approach. I'll check and comeback on this.

Reply:
> For example, if we lose the local data in all replicas, the lastStableOffset could still be in the middle of a tiered segment and moving it to localLogStartOffset immediately will be incorrect.

I'm not clear on this:

1. Segments that are eligible for upload to remote storage only when the `lastStableOffset` moves beyond the segment-to-be-uploaded-end-offset. 
2. When all the replicas loses local data (offline partition), then we consider the data in remote storage also lost. Currently, for this case, we don't have provision to serve the remote data.
3. When `firstUnstableOffsetMetadata` is empty, we return `highWatermark`. With this patch, the `highWatermark` lower boundary is set to `localLogStartOffset` so there won't be an issue. 

> Note that OffsetMetadata (segmentBaseOffset and relativePositionInSegment) is only used in DelayedFetch for estimating the amount of available bytes.

The [LogOffsetMetadata#onOlderSegment](https://sourcegraph.com/github.com/apache/kafka@5de5d967adffd864bad3ec729760a430253abf38/-/blob/storage/src/main/java/org/apache/kafka/storage/internals/log/LogOffsetMetadata.java?L54) method is used in the [hot-path](https://sourcegraph.com/github.com/apache/kafka@5de5d967adffd864bad3ec729760a430253abf38/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L324) of incrementing the high-watermark and expects the full metadata, otherwise it throws an error. Is it ok to remove the throwable from LogOffsetMetadata#onOlderSegment method and return `false` by default.

Reply:
> I'm not clear on this:
> 
> 1. Segments that are eligible for upload to remote storage only when the lastStableOffset moves beyond the segment-to-be-uploaded-end-offset.
> 2. When all the replicas loses local data (offline partition), then we consider the data in remote storage also lost. Currently, for this case, we don't have provision to serve the remote data.
> 3. When firstUnstableOffsetMetadata is empty, we return highWatermark. With this patch, the highWatermark lower boundary is set to localLogStartOffset so there won't be an issue.
> 

That's true. It's just that that is yet another offset that we need to bound. I am also not sure if there are other side effects of adjusting HWM and LSO.

Left some comments on https://github.com/apache/kafka/pull/15825.